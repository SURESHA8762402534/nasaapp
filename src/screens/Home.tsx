/** @format */

import { Snackbar } from "@material-ui/core";
import { Close } from "@material-ui/icons";
import { useNavigation } from "@react-navigation/native";
import { NativeStackNavigationProp } from "@react-navigation/native-stack";
import axios from "axios";
import React, { FC, useState } from "react";
import {
  ActivityIndicator,
  Alert,
  Button,
  SafeAreaView,
  TextInput,
  View,
} from "react-native";
import { styles } from "../styles/screens";
import { KEY } from "../types/screens";

const Home: FC = () => {
  const [text, setText] = useState<string>("");
  const [loading, setLoading] = useState<boolean>(false);
  const [error, setError] = useState<boolean>(false);
  const [state, setState] = React.useState<any>({
    vertical: "center",
    horizontal: "center",
  });
  const { vertical, horizontal, open } = state;

  const action = <Close onClick={() => setError(false)} />;

  const navigation = useNavigation<NativeStackNavigationProp<any>>();

  const getRandomAsteroid = () => {
    setLoading(true);
    axios
      .get("https://api.nasa.gov/neo/rest/v1/neo/browse?api_key=DEMO_KEY")
      .then((res) => {
        const randomNumber = Math.floor(
          Math.random() * res.data.near_earth_objects.length
        );
        getdetailes(res.data.near_earth_objects[randomNumber].id);
      })
      .catch(() => {
        Alert.alert("Error", "Something Went wrong", [
          {
            text: "ok",
          },
        ]);
      })
      .finally(() => setLoading(false));
  };

  const getdetailes = async (id: any) => {
    setLoading(true);
    axios
      .get(`https://api.nasa.gov/neo/rest/v1/neo/${id}?api_key=${KEY}`)
      .then((res) => {
        navigation.navigate("`", { data: res.data });
      })
      .catch(() => {
        setError(true);
      })
      .finally(() => {
        setLoading(false);
      });
  };

  return (
    <SafeAreaView testID="home" style={styles.container}>
      <Snackbar
        anchorOrigin={{ vertical, horizontal }}
        open={error}
        onClose={() => setError(false)}
        message="Wrong Id"
        key={vertical + horizontal}
        action={action}
        autoHideDuration={2000}
      />
      <View style={styles.inputcontainer}>
        <TextInput
          testID="input"
          placeholder="Enter Asteroid ID"
          placeholderTextColor={"grey"}
          onChangeText={(text) => setText(text)}
          style={styles.input}
        />

        <View style={styles.btn}>
          <Button
            testID="btn"
            disabled={text.length == 0}
            title="Submit"
            onPress={() => {
              getdetailes(text);
            }}
            color="orange"
          />
        </View>
      </View>
      <View>
        {loading ? (
          <View style={styles.load}>
            <ActivityIndicator color={"blue"} size={20} />
          </View>
        ) : (
          <View style={styles.buttonsRow}>
            <Button
              testID="random"
              disabled={loading}
              title="Random Asteroid"
              onPress={() => {
                getRandomAsteroid();
              }}
              color="orange"
            />
          </View>
        )}
      </View>
    </SafeAreaView>
  );
};

export default Home;
