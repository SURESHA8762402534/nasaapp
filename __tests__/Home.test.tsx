import "@testing-library/jest-native/extend-expect";
import "@testing-library/jest-native";
import { fireEvent, render } from "@testing-library/react-native";
import React from "react";
import Home from "../src/screens/Home";

jest.mock("@react-navigation/native", () => {
  return {
    useNavigation: () => ({
      navigate: jest.fn(),
    }),
  };
});

describe("Home", () => {
  it("renders without crash", () => {
    render(<Home />);
  });

  it("matches snapshot", () => {
    const tree = render(<Home />);
    expect(tree.toJSON()).toMatchSnapshot();
  });

  it("has component with req testID", () => {
    const tree = render(<Home />);
    expect(tree.getByTestId("home")).toBeTruthy();
    const txt = tree.getByTestId("input");
    const btn = tree.getByTestId("btn");
    const randombtn = tree.getByTestId("random");
    expect(btn).toBeDisabled();
    fireEvent.press(randombtn);

    fireEvent.changeText(txt, { target: { value: "test" } });
    fireEvent.press(btn);
  });
});
