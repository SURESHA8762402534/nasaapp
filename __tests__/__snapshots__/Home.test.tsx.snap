// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home matches snapshot 1`] = `
<RCTSafeAreaView
  emulateUnlessSupported={true}
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#f6f1eb",
      "flex": 1,
      "justifyContent": "center",
      "padding": 10,
      "width": "100%",
    }
  }
  testID="home"
>
  <View
    style={
      Object {
        "alignItems": "center",
        "borderColor": "black",
        "borderRadius": 10,
        "borderStyle": "solid",
        "borderWidth": 1,
        "display": "flex",
        "flexDirection": "row",
        "justifyContent": "center",
        "paddingBottom": 10,
        "paddingTop": 10,
        "width": "90%",
      }
    }
  >
    <TextInput
      allowFontScaling={true}
      onChangeText={[Function]}
      placeholder="Enter Asteroid ID"
      placeholderTextColor="grey"
      rejectResponderTermination={true}
      style={
        Object {
          "backgroundColor": "white",
          "borderColor": "blue",
          "borderRadius": 10,
          "borderWidth": 1,
          "marginBottom": 10,
          "padding": 10,
          "width": "80%",
        }
      }
      testID="input"
      underlineColorAndroid="transparent"
    />
    <View
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "orange",
          "borderRadius": 10,
          "borderWidth": 1,
          "bottom": 5,
          "display": "flex",
          "justifyContent": "center",
          "marginLeft": 5,
          "position": "relative",
        }
      }
    >
      <View
        accessibilityRole="button"
        accessibilityState={
          Object {
            "disabled": true,
          }
        }
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "opacity": 1,
          }
        }
        testID="btn"
      >
        <View
          style={
            Array [
              Object {},
              Object {},
            ]
          }
        >
          <Text
            disabled={true}
            style={
              Array [
                Object {
                  "color": "#007AFF",
                  "fontSize": 18,
                  "margin": 8,
                  "textAlign": "center",
                },
                Object {
                  "color": "orange",
                },
                Object {
                  "color": "#cdcdcd",
                },
              ]
            }
          >
            Submit
          </Text>
        </View>
      </View>
    </View>
  </View>
  <View>
    <View
      style={
        Object {
          "marginTop": 30,
        }
      }
    >
      <View
        accessibilityRole="button"
        accessibilityState={Object {}}
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "opacity": 1,
          }
        }
        testID="random"
      >
        <View
          style={
            Array [
              Object {},
            ]
          }
        >
          <Text
            disabled={false}
            style={
              Array [
                Object {
                  "color": "#007AFF",
                  "fontSize": 18,
                  "margin": 8,
                  "textAlign": "center",
                },
                Object {
                  "color": "orange",
                },
              ]
            }
          >
            Random Asteroid
          </Text>
        </View>
      </View>
    </View>
  </View>
</RCTSafeAreaView>
`;
